src/cluster.c

```diff
@@ -2423,19 +2438,26 @@ void clusterBuildMessageHdr(clusterMsg *hdr, int type) {
         hdr->myip[NET_IP_STR_LEN-1] = '\0';
     }
 
-    /* Handle cluster-announce-port as well. */
+    /* Handle cluster-announce-[tls-|bus-]port. */
     int port = server.tls_cluster ? server.tls_port : server.port;
-    int announced_port = server.cluster_announce_port ?
-                         server.cluster_announce_port : port;
-    int announced_cport = server.cluster_announce_bus_port ?
-                          server.cluster_announce_bus_port :
-                          (port + CLUSTER_PORT_INCR);
+    int announced_port = port;
+    int announced_pport = server.tls_cluster ? server.port : 0;
+    int announced_cport = port+CLUSTER_PORT_INCR;
+    if (server.tls_cluster && server.cluster_announce_tls_port) {
+        announced_port = server.cluster_announce_tls_port;
+        announced_pport = server.cluster_announce_port;
+    } else if (server.cluster_announce_port) {
+        announced_port = server.cluster_announce_port;
+    }
+    if (server.cluster_announce_bus_port)
+        announced_cport = server.cluster_announce_bus_port;
```

Comment:
small nitpick: the logic of deriving the announced port/pport is duplicate, consider moving it to a single helper func.

Reply:
Good idea. Done.